shachain/store.go

```diff
@@ -0,0 +1,187 @@
+// by given indexes.
+//
+// Description: The Lightning Network wants a chain of (say 1 million)
+// unguessable 256 bit values; we generate them and send them one at a time
+// to a remote node.  We don't want the remote node to have to store all the
+// values, so it's better if they can derive them once they see them.
+type SecretStore interface {
+	// Restore function is ised to restore the previous secret by its index.
+	Restore(uint64) (*wire.ShaHash, error)
+
+	// Store is used to store the given sha hash in efficient maner.
+	Store(*wire.ShaHash) error
+
+	// ToBytes convert store to the binary representation.
+	ToBytes() ([]byte, error)
+}
+
+// ShaChainStore implementation of SecretStore. This version of shachain store
+// sligtly changed in terms of method naming. Initial concept might be found
+// here:
+// https://github.com/rustyrussell/ccan/blob/master/ccan/crypto/shachain/design.txt
+type ShaChainStore struct {
+	// lenBuckets stores the number of currently active buckets.
+	lenBuckets uint8
+
+	// buckets is an array of elements from which we may derive all previous
+	// elements, each bucket corresponds to the element index number of
+	// trailing zeros.
+	buckets [maxHeight]*Element
+
+	// index is an available index which will be assigned to the new
+	// element.
+	index index
+}
+
+// A compile time check to ensure ShaChainStore implements the SecretStore
+// interface.
+var _ SecretStore = (*ShaChainStore)(nil)
+
+// NewShaChainStore creates the new shachain store.
+func NewShaChainStore() *ShaChainStore {
+	return &ShaChainStore{
+		lenBuckets: 0,
+		index:      startIndex,
+	}
+}
+
+// NewShaChainStoreFromBytes recreates the initial store state from the given
+// binary shachain store respresentation.
+func NewShaChainStoreFromBytes(data []byte) (*ShaChainStore, error) {
+	var err error
+
+	receiver := &ShaChainStore{}
+	buf := bytes.NewBuffer(data)
+
+	err = binary.Read(buf, binary.BigEndian, &receiver.lenBuckets)
+	if err != nil {
+		return nil, err
+	}
+
+	i := uint8(0)
+	for ; i < receiver.lenBuckets; i++ {
+		element := &Element{}
+
+		err = binary.Read(buf, binary.BigEndian, &element.index)
+		if err != nil {
+			return nil, err
+		}
+
+		hash, err := wire.NewShaHash(buf.Next(wire.HashSize))
+		if err != nil {
+			return nil, err
+		}
+		element.hash = hash
+		receiver.buckets[i] = element
+	}
+
+	err = binary.Read(buf, binary.BigEndian, &receiver.index)
+	if err != nil {
+		return nil, err
+	}
+
+	return receiver, nil
+}
+
+// Restore function is used to restore the previous secret by its index. If
+// secret which correponds to given index was not previously placed in store we
+// will not able to derive it and function will fail.
+//
+// NOTE: This function is part of the SecretStore interface.
+func (store *ShaChainStore) Restore(v uint64) (*wire.ShaHash, error) {
+	ind := NewIndex(v)
+
+	// Trying to derive the index from one of the existing buckets elements.
+	i := uint8(0)
+	for ; i < store.lenBuckets; i++ {
```

Comment:
Any reason why you didn't go with the full 3-clause for-loop here? It would eliminate the extra line above.